include/RevThread.h

```diff
@@ -0,0 +1,200 @@
+  START,    // Initial state; RevCPU has not yet registered it in its Threads table
+  RUNNING,  // Thread is assigned and executing on a Proc/HART (NOTE: This does NOT change if a Thread is stalled)
+  BLOCKED,  // Waiting for thread synchronization at this point (Currently only triggered by call to `rev_pthread_join`)
+  READY,    // In the ready list; waiting for CPU (not implemented yet).
+  DONE,     // Thread has finished; deallocate resources.
+};
+
+class RevThread {
+ private:
+  uint32_t ThreadID;                                   // Thread ID
+  uint32_t ParentThreadID;                             // Parent ThreadID
+  uint64_t StackPtr;                                   // Initial stack pointer
+  uint64_t FirstPC;                                    // Initial PC
+  std::shared_ptr<RevMem::MemSegment> ThreadMem;       // TLS and Stack memory
+  RevFeature* Feature;                                 // Feature set for this thread
+  uint64_t ThreadPtr;                                  // Thread pointer
+  ThreadState State = ThreadState::START;              // Thread state (Initializes as )
+  RevRegFile RegFile;                                  // Register file
+  std::unordered_set<uint32_t> ChildrenThreadIDs = {}; // Child thread IDs
+  std::unordered_set<int> fildes = {0, 1, 2  };      // Default file descriptors
+
+  uint32_t WaitingToJoinTID = __INVALID_TID__;
+
+ public:
+  RevThread(uint32_t inputThreadID, uint32_t inputParentThreadID,
+            uint64_t inputStackPtr, uint64_t inputFirstPC,
+            std::shared_ptr<RevMem::MemSegment>& inputThreadMem,
+            RevFeature* inputFeature)
+   : ThreadID(inputThreadID), ParentThreadID(inputParentThreadID), StackPtr(inputStackPtr),  
+     FirstPC(inputFirstPC), ThreadMem(inputThreadMem), Feature(inputFeature) {
+  // Create the RegFile for this thread
+  RegFile = RevRegFile(); 
+  
+  // Set the stack pointer 
+  RegFile.SetX(Feature, 2, StackPtr );
+
+  // Set the thread pointer
+  ThreadPtr = ThreadMem->getTopAddr();
+  RegFile.SetX(Feature, 2, ThreadPtr);
+
+  // Set the PC 
+  RegFile.SetPC(Feature, FirstPC);
+}
+
+  // ThreadID operations
+  uint32_t GetThreadID() const { return ThreadID; }
+  void SetThreadID(const uint32_t NewThreadID) { ThreadID = NewThreadID; }
+  uint32_t GetParentThreadID() const { return ParentThreadID; }
+  void SetParentThreadID(const uint32_t NewParentThreadID) { ParentThreadID = NewParentThreadID; }
+  uint32_t GetWaitingToJoinTID() const { return WaitingToJoinTID; }
+  void SetWaitingToJoinTID(const uint32_t ThreadToWaitOn) { WaitingToJoinTID = ThreadToWaitOn; }
+
+  // Add new file descriptor 
+  void AddFD(int fd){ fildes.insert(fd); }
+
+  // Remove file descriptor from this thread (ie. rev_close)
+  void RemoveFD(int fd){ fildes.erase(fd); }
+
+  // See if file descriptor exists/is owned by this thread
+  bool FindFD(int fd){ return fildes.count(fd); }
+
+  // Get the threads 
+  std::unordered_set<int> GetFildes(){ return fildes; }
```

Comment:
This `GetFildes` function is not used (the comment is confusing). If used, you should consider:
```c++
 const std::unordered_set<int>& GetFildes(){ return fildes; }
```
... to prevent expensive copying of the entire container.